wolfcrypt/src/port/iotsafe/iotsafe.c

```diff
@@ -0,0 +1,1064 @@
+    if (iotsafe->enabled) {
+        ret = iotsafe_gen_keypair(iotsafe->privkey_id);
+        if (ret < 0)
+            return ret;
+        ret = iotsafe_get_public_key(iotsafe->privkey_id, key);
+    } else {
+        WC_RNG    rng;
+        ret = wc_InitRng(&rng);
+        if (ret != 0)
+            return ret;
+        ret = wc_ecc_init(key);
+        if (ret == 0) {
+            /* create new key */
+            ret = wc_ecc_make_key_ex(&rng, keySz, key, ecc_curve);
+        }
+        wc_FreeRng(&rng);
+    }
+    return ret;
+}
+
+
+static int wolfIoT_ecc_sign(WOLFSSL* ssl,
+       const unsigned char* in, unsigned int inSz,
+       unsigned char* out, word32* outSz,
+       const unsigned char* keyDer, unsigned int keySz,
+       void* ctx)
+{
+    int ret;
+    (void)ctx;
+    IOTSAFE *iotsafe = &ssl->iotsafe;
+    WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_sign\n");
+
+    if (iotsafe->enabled) {
+        ret = iotsafe_sign_hash(iotsafe->privkey_id, IOTSAFE_HASH_SHA256, IOTSAFE_SIGN_ECDSA,
+                in, inSz, out, outSz);
+        if (ret > 0)
+            return 0;
+        else
+            return ret;
+    } else {
+        WC_RNG    rng;
+        word32    idx = 0;
+        ecc_key   myKey;
+        byte*     keyBuf = (byte*)keyDer;
+        ret = wc_InitRng(&rng);
+        if (ret != 0)
+            return ret;
+
+        ret = wc_ecc_init(&myKey);
+        if (ret == 0) {
+            ret = wc_EccPrivateKeyDecode(keyBuf, &idx, &myKey, keySz);
+            if (ret == 0) {
+                ret = wc_ecc_sign_hash(in, inSz, out, outSz, &rng, &myKey);
+            }
+            wc_ecc_free(&myKey);
+        }
+        wc_FreeRng(&rng);
+    }
+    return ret;
+}
+
+static int wolfIoT_ecc_verify(WOLFSSL *ssl,
+       const unsigned char* sig, unsigned int sigSz,
+       const unsigned char* hash, unsigned int hashSz,
+       const unsigned char* keyDer, unsigned int keySz,
+       int* result, void* ctx)
+{
+    byte sig_raw[IOTSAFE_ECC_KSIZE* 2];
+    int ret;
+    ecc_key key;
+    word32 r_size = IOTSAFE_ECC_KSIZE, s_size = IOTSAFE_ECC_KSIZE;
+    word32 inOutIdx = 0;
+    IOTSAFE *iot = &ssl->iotsafe;
+    byte pubkey_slot = iot->peer_cert_slot;
+
+    (void)ctx;
+
+    if (iot->enabled) {
+        /* Invalidate verification, by default. */
+        *result = 0;
+        WOLFSSL_MSG("IOTSAFE: Called wolfIoT_ecc_verify\n");
+
+        /* Convert ECC signature into R,S */
+        wc_ecc_sig_to_rs(sig, sigSz, sig_raw, &r_size, sig_raw
+                + IOTSAFE_ECC_KSIZE, &s_size);
+        /* Store the server's public key in IoT-safe vault */
+
+        /* Create an ecc object to handle the key */
+        ret = wc_ecc_init(&key);
+        if (ret != 0)
+            return ret;
+
+        /* Import from keyDer argument */
+        ret = wc_EccPublicKeyDecode(keyDer, &inOutIdx, &key, keySz);
+        if (ret < 0) {
+            WOLFSSL_MSG("Could not decode public key");
+            return ret;
```

Comment:
Needs to call `wc_ecc_free`. Same with other returns below. Perhaps refactor for single return point to simplify?

Reply:
Refactored the three functions to have single point of return after wc_ecc_init, with wc_ecc_free called in all cases.